core/unit_test/incremental/Test03a_ParallelFor.hpp

```diff
@@ -0,0 +1,175 @@
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+/// @Kokkos_Feature_Level_Required:3
+// parallel-for unit test.
+// In this test different elements of an array are updated by different threads.
+
+// A macro representing the values assigned to the elements of the array.
+// This macro is used while updating the array and while checking the
+// correctness.
+#define updateValue (i + 1) * value
+
+namespace Test {
+
+using value_type       = double;
+int num_elements       = 10;
+const value_type value = 0.5;
+
+struct ParallelForFunctor {
+  value_type *_data;
+
+  ParallelForFunctor(value_type *data) : _data(data) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int i) const { _data[i] = updateValue; }
+};
+
+template <class ExecSpace>
+struct TestParallel_For {
+  // Memory space type for Device and Host data
+  using d_memspace_type = typename ExecSpace::memory_space;
+  using h_memspace_type = Kokkos::HostSpace;
+
+  value_type *deviceData, *hostData;
+
+  // Check if the array values are updated correctly
+  void correctness_check(value_type *data) {
+    for (int i = 0; i < num_elements; ++i) {
+      ASSERT_EQ(data[i], updateValue)
+          << "Values in index " << i << " are incorrect";
+    }
+  }
+
+  // Routine to allocate memory in a specific memory space.
+  template <class MemSpace>
+  value_type *allocate_mem(int N) {
+    return (static_cast<value_type *>(
+        Kokkos::kokkos_malloc<MemSpace>("deviceData", N * sizeof(value_type))));
+  }
```

Comment:
We could consider moving these into `Kokkos` (but certainly not in this pull request).

Reply:
What do you mean? What exactly do you want which is different from kokkos_malloc?

Reply:
Something like a version templated on the type to be allocated, such that we can avoid casting the resulting pointer and can pass the number of objects instead of the size in bytes.